fuel-tx/src/transaction/fee.rs

```diff
@@ -113,18 +120,23 @@ impl TransactionFee {
     /// Attempt to create a transaction fee from parameters and transaction internals
     ///
     /// Will return `None` if arithmetic overflow occurs.
-    pub fn checked_from_tx<T: Chargeable>(
+    pub fn checked_from_tx<T: Chargeable + Inputs>(
+        gas_costs: &GasCosts,
         params: &FeeParameters,
         tx: &T,
     ) -> Option<Self> {
         let metered_bytes = tx.metered_bytes_size() as Word;
+        let unique_witnesses = tx.signed_inputs_with_unique_witnesses().len() as Word;
```

Comment:
Because you need to find `Signed` and `Predicate` inputs, it is better to iterate over them here and calculate the number of unique witnesses, plus charge for predicates.

You can use one loop to cover all inputs. Plus, you can already implement charging for the predicates too, but with some dummy values that you will update later.

Reply:
I will point out that the calculation for predicate gas is done by the `tx`:
```
tx.gas_used_by_predicates()
```
whereas the signature cost is done here.

Would it be wrong to have a
```
tx.gas_used_for_sign_verification()
```
method on the `tx`? And if yes, would it make sense to change the predicate calculation?

I don't really have an opinion yet, but seems strange that they are so different.

Reply:
I think it may be beneficial to move this to the `Chargeable` trait, and allow the method to take in a reference to `GasCosts`. We calculate the gas cost when checking the transaction, but also in a number of tests to verify the fee. Ideally, we want to test the calculation used in production, rather than testing a recreation of the calculation that may become out of sync. This means we need to wrap it in a method that we can call on the tx.